### DISCUSSION TITLE
socket close when running as client in node

### DISCUSSION THREAD

**Discussion by jkristia** [enisdenjo/graphql-ws#327] [Q&A] [answered]
This is more of a question.

I'm trying to use the library as a client in node / typescript.  
I have followed the recipe for 'client usage in node', but I am missing something as I get 'socket closed' error immediately.

```typescript
// https://github.com/enisdenjo/graphql-ws
import ws from 'ws';
import { createClient } from 'graphql-ws';
import { randomUUID } from 'crypto';

const client = createClient({
	url: 'ws://10.36.83.233:5513/graphql',
	webSocketImpl: ws,
	generateID: () => randomUUID()
});

export const graphql_play = () => {
	const sub = client.subscribe(
		{
			query: 'subscription { Resource { mutation instance { id }}}'
		},
		{
			next: (data) => console.log('data"', data),
			error: (error) => { console.log('error:', error) },
			complete: () => console.log('finished')
		}
	)
}
```
main from where it is called
```typescript
import { graphql_play} from './graphql-play'

function wait(): Promise<any> {
    return new Promise( () => {});
}
async function run() {
    console.log('run')
    graphql_play();
    await wait();
    console.log('end')
}

run();
```

The returned error is 

```
error
CloseEvent {Symbol(kTarget): WebSocket, Symbol(kType): 'close', Symbol(kCode): 1002, Symbol(kReason): '', Symbol(kWasClean): true}
code: 1002
reason: ''
Symbol(kCode): 1002
Symbol(kReason): ''
Symbol(kTarget): WebSocket {_events: {…}, _eventsCount: 4, _maxListeners: undefined, _binaryType: 'nodebuffer', _closeCode: 1002, …}
Symbol(kType): 'close'
Symbol(kWasClean): true
target: WebSocket
type: 'close'
wasClean: true
[[Prototype]]: Event
error: CloseEvent {Symbol(kTarget): WebSocket, Symbol(kType): 'close', Symbol(kCode): 1002, Symbol(kReason): '', Symbol(kWasClean): true}

```

Same work if I use the deprecated `subscription-transport-ws` following this [boilerplate example](https://github.com/hasura/nodejs-graphql-subscriptions-boilerplate)

**Comment by enisdenjo** [comment]
It is very likely that your server does not support `graphql-ws`.

Note that the [deprecated `subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws) and `graphql-ws` are not cross-compatible, [as stated in the readme disclaimer](https://github.com/enisdenjo/graphql-ws#disclaimer).

**Comment by jkristia** [comment, marked as answer]
I appreciate you quick response, and I did miss the disclaimer part.
I think you are right about incompatibility issue.
The service I'm trying to connect to is running `"apollo-server-express": "^2.14.0",`
I will continue using the deprecated package until the serve is upgraded.

Appreciate your help.

**Reply by enisdenjo** [reply]
If you have control over the server, you can use the ["ws server usage with Apollo Server Express" recipe](https://github.com/enisdenjo/graphql-ws#apollo-server-express) to add `graphql-ws` support to `apollo-server-express`.

**Comment by jkristia** [comment]
yeah, unfortunately I do not have control of that service. But I have requested for an update. Again, thank you for your help
